decisions/DR012-Entity-reference-validation-workflow.md

```diff
@@ -0,0 +1,355 @@
+have been validated.
+
+### References
+
+See issue [#532](https://github.com/OpenAssetIO/OpenAssetIO/issues/532).
+
+## Options considered
+
+### Option 1
+
+Allow hosts to call `isEntityReference` as well as
+`createEntityReference` (which throws if given an invalid reference).
+
+#### Pros
+
+- No need to (try to) construct an object just for a boolean comparison.
+- Obvious self-documenting usage.
+- Clear language bindings based on existing patterns.
+- No way to construct an invalid entity reference.
+
+#### Cons
+
+- Duplicated validation effort if the code path goes on to use
+  `createEntityReference`.
+- Hosts may forget to check `isEntityReference` first, leading to
+  unexpected exceptions.
+
+### Option 2
+
+`isEntityReference` is hidden from the host. `createEntityReference` is
+the only way to construct and validate an entity reference, and
+throws if invalid.
+
+#### Pros
+
+- Only one way to accomplish validation and construction.
+- Validation and construction are performed in one operation.
+- Clear language bindings based on existing patterns.
+- No way to construct an invalid entity reference.
+
+#### Cons
+
+- Host must use exceptions to test validity, increasing boilerplate
+  and reducing performance.
+
+### Option 3
+
+`isEntityReference` hidden from the host. `createEntityReferenceIfValid`
+is instead used to return an `optional<EntityRef>` in C++
+(`None`/`EntityRef` in Python; `NULL` (and error code)/`EntityRef_h` in
+C).
+
+#### Pros
+
+- Only one way to accomplish validation and construction.
+- Validation and construction are performed in one operation.
+- Lowest host boilerplate for many workflows, utilising modern C++.
+- If using  `optional::value()` to extract the `EntityRef` before
+  passing to API methods, then safety is guaranteed, since a
+  `bad_optional_access` will be thrown if invalid (though less
+  informative than a custom exception).
+- Pybind has built-in support for `std::optional`, translating "not set"
+  to `None`.
+
+#### Cons
+
+- Very slightly less performant than `isEntityReference` in the invalid
+  reference case (at the risk of second-guessing compiler
+  optimisations).
+- Not immediately obvious in Python that the host must test the returned
+  value is valid (not `None`).
+- Uses a different optional API to `TraitsData`, decided
+  in [DR010](DR010-TraitsData-value-retrieval-API.md).
+  Though many of the arguments do not hold for this case (e.g.
+  direct-assignment to existing variables; detailed result codes).
+- Using the dereferencing operator (`*`) on a `std::optional` is unsafe
+  and loses the compile-time guarantee that the entity reference is
+  safe before being passed to API methods.
+- Language bindings to C and Python have a different interface than C++.
+
+## Outcome
+
+Based on the sketches in the Appendix (below) and the various pros and
+cons we have decided on supporting all workflows, i.e. implementing
+both Option 1/2 and 3.
```

Comment:
[consider]
```suggestion
both Option 1, 2 and 3.
```

Reply:
I wasn't sure about this, since option 1 contains option 2. Maybe we should just get rid of the `2`.

Reply:
Oh right got you sorry, leave as is if you like.

Reply:
Updated the wording in 5927528c, see what you think.

Reply:
LGTM!